pkg/upgrade/upgrade.go

```diff
@@ -65,6 +94,66 @@ func TryUpgrade(cli *clientv3.Client) error {
 
 // upgradeToVer1 does upgrade operations from Ver0 to Ver1.
 // in fact, this do nothing now, and just for demonstration.
-func upgradeToVer1(cli *clientv3.Client) error {
+func upgradeToVer1(cli *clientv3.Client, uctx Context) error {
+	return nil
+}
+
+// upgradeToVer2 does upgrade operations from Ver1 to Ver2 (v2.0.0-rc3) to upgrade syncer checkpoint schema
+func upgradeToVer2(cli *clientv3.Client, uctx Context) error {
+	upgradeTaskName := "upgradeToVer2"
+	logger := log.L().WithFields(zap.String("task", upgradeTaskName))
+
+	if uctx.SubTaskConfigs == nil {
+		logger.Info("no downstream DB, skipping")
+		return nil
+	}
+
+	// tableName -> DBConfig
+	dbConfigs := map[string]config.DBConfig{}
+	for task, m := range uctx.SubTaskConfigs {
+		for sourceID, subCfg := range m {
+			tableName := dbutil.TableName(subCfg.MetaSchema, cputil.SyncerCheckpoint(subCfg.Name))
+			subCfg2, err := subCfg.DecryptPassword()
+			if err != nil {
+				log.L().Error("skip subconfig when upgrading", zap.String("task", task),
+					zap.String("source id", sourceID), zap.Error(err))
+			}
+			dbConfigs[tableName] = subCfg2.To
+		}
+	}
+
+	toClose := make([]*conn.BaseDB, 0, len(dbConfigs))
+	defer func() {
+		for _, db := range toClose {
+			db.Close()
+		}
+	}()
+	for tableName, cfg := range dbConfigs {
+		targetDB, err := conn.DefaultDBProvider.Apply(cfg)
+		if err != nil {
+			logger.Error("skip target DB when upgrading", zap.String("table name", tableName))
+			return err
```

Comment:
`return` rather than `continue`? if so, how about `return` after L119? if `continue`, log `err` in L134.

Reply:
choose `return`

https://github.com/pingcap/dm/pull/988/commits/fa401af29be9b3bbb7bf5e3cfd6bd7d96f78c800